---
title: "RiboR"
author: "Michael Geng"
output:
  html_document:
    theme: lumen
    toc: true
    toc_depth: 4
    toc_float: true
  pdf_document:
    toc: true
    toc_depth: 4
  rmarkdown:: html_vignette
vignette: >
  %\VignetteIndexEntry{RiboR} 
  %\VignetteEngine{knitr::rmarkdown} 
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This package is an R interface for .ribo files generated by the riboflow pipeline and extends the workflow of the pipeline into an R environment. The package offers a suite of reading functions for any of the datasets present in a .ribo file and also provides some rudimentary plotting functions.

This document provides a short demonstration of the RiboR package.
More detailed documentation and examples can be found in the github repository below.

## Installation 
The source code of RiboR package is in a [public Github repository](https://github.com/ribosomeprofiling/ribor).

### Requirements
* **R:** RiboR requires **version 3.6 or later**. Previous R versions are likely to cause installation or run time problems.

* **Devtools:** Installing the package requires `devtools`, a CRAN package that provides a wide suite of development tools in the R environment. 

To install the package, run the following lines of code.
```{r, eval = FALSE}
install.packages("devtools")
library("devtools")
install_github("ribosomeprofiling/ribo/ribor")
```

## Getting Started
``` {r, eval = TRUE}
library(ribor)
```

In order to interact with the .ribo file, it is necessary to create a ribo object that provides a direct handle to the file and displays its various attributes.

```{r createRibo}
#file path to the example ribo file 
file.path <- system.file("extdata", "HEK293_ingolia.ribo", package = "ribor")

#generates the 'ribo' class object 
original.ribo <- create_ribo(file.path , rename = rename_default)
```

We can take a quick look at the contentsd of the `original.ribo` object.

```{r}
original.ribo
```

We explore this ribo file in this document. We start with generating plots.

## Visualising Data

### Length Distribution

We can see the distribution of the reads according to their lengths using the function `plot_length_distribution`. 

``` {r plot_length_distribution ribo, fig.width = 7}
plot_length_distribution(x           = original.ribo,
                         region      = "CDS",
                         range.lower = 28,
                         range.upper = 32,
                         fraction    = TRUE)
```

In the example above, the y-axis contains the fraction of the reads for easy comparison. 

### Metagene

Metagene plots for start and stop sites can be generated using the function `plot_metagene`.

```{r plot_metagene ribo, fig.width = 7}
plot_metagene(original.ribo,
              site        = "start",
              normalize   = FALSE,
              experiments = c("GSM1606108"),
              title       = "Start Site Coverage",
              range.lower = 28,
              range.upper = 32)
```

To plot all experiments in the ribo file, simply remove `experiments` from the arguments. For easy comparison, we also turn on normalization by setting
`normalize = TRUE,`

```{r plot_metagene ribo_all_exps, fig.width = 7}
plot_metagene(original.ribo,
              site        = "stop",
              normalize   = TRUE,
              title       = "Stop Site Coverage",
              range.lower = 28,
              range.upper = 32)
```

### Region Counts
The following code will plot the region counts of the regions UTR5, CDS, and UTR3 as a stacked bar plot. 

``` {r plot_region_counts ribo, fig.width = 7}
plot_region_counts(x           = original.ribo,
                   range.lower = 28,
                   range.upper = 32)
```

## Retrieving Data 

### Metagene Data

A very typic al way of looking at metagene data is summing values accrsoss diffferent
lengths and transcripts. This is done by setting `length = TRUE` and `transcript = TRUE`. In the following example, we obtain the metagene data for the start site.

```{r}
#get the start site across read lengths 28 to 32
meta.start <- get_metagene(ribo.object = original.ribo, 
                           site        = "start",
                           range.lower = 28,
                           range.upper = 32,
                           length      = TRUE,
                           transcript  = TRUE)
#only print first ten columns 
print(meta.start[ , 1:10])
```

We can also obtain the same metagene data in tidy form in a data table.

```{r}
tidy.meta.start <- get_tidy_metagene(ribo.object = original.ribo,
                                     site        = "start",
                                     range.lower = 28,
                                     range.upper = 32,
                                     length      = TRUE)
head(tidy.meta.start)
```

### Region Counts 

The following code gives us the number of reads mapped to the Coding Sequence (CDS) region of the transcripts for the ribosome protected footprints of length 30 nucleotides.

``` {r get_region_counts transcript}
rc <- get_region_counts(original.ribo,
                        range.lower = 30,
                        range.upper = 30,
                        length      = TRUE,
                        transcript  = FALSE,
                        alias       = TRUE,
                        region      = c("CDS"))
rc
```

### Transcript Coverage

Transcript coverage is an optional field of ribo file. When it exists,
it can be read into a data frame using the `get_coverage` function.
Transcript coverage can be obtained for one transcript at a time. Here we obtain the coverage of MYC_206.

```{r}
cov <- get_coverage(ribo.object = original.ribo,
                    name        = "MYC-206",
                    range.lower = 28,
                    range.upper = 32,
                    length      = TRUE,
                    alias       = TRUE,
                    tidy        = TRUE,
                    experiments = c("GSM1606107"))
cov[380:390]
```

### RNA-Seq

RNA-Seq is another optional field of the ribo file.
We can extract the RNA-Seq data of the experiment GSM1606107
 for the regions UTR5, CDS and UTR3 as follows.

```{r}
experiment.info <- get_info(ribo.object = original.ribo)[['experiment.info']]
has.rnaseq <- experiment.info[experiment.info$rna.seq == TRUE, experiment]
 
rnaseq <- get_rnaseq(ribo.object = original.ribo,
                     tidy        = TRUE,
                     alias       = TRUE,
                     regions     = c("UTR5J", "CDS", "UTR3J"),
                     experiments = c("GSM1606107"))
head(rnaseq)
```

## Metadata

Metadata, for the entire ribo file and for the individual experiments, are
stored in ribo files if they were provided in file creation.

To see the metadata of the ribo file, we use the function `get_metadata`.

```{r get_metadata file, eval = FALSE}
get_metadata(ribo.object = original.ribo, 
             print = TRUE)
#The output for this is quite long, so here is a truncated version.
```

```{r get_metadata shortened_meta, echo = FALSE}
info <- capture.output(get_metadata(ribo.object = original.ribo, 
                                    print = TRUE))
cat(info[1:15], sep = "\n")
```

Let's recall the contents of the ribo file. 

```{r}
original.ribo
```

According to the summary above, 
two experiments GSM1606107 and GSM1606108 have metadata. Below, we look at the metadata of the experiment GSM1606107.

```{r get_metadata_exp}
get_metadata(ribo.object = original.ribo,
             name        = "GSM1606107",
             print       = TRUE)

```
